File: examples/python/generate_video.py
"""
Generate a video using Sats4AI L402 API (async service).

Usage:
    python generate_video.py "A robot dancing in Times Square"

Requires: pip install requests
"""

import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../.."))
from lib.l402_client import L402Client

prompt = sys.argv[1] if len(sys.argv) > 1 else "A robot dancing in Times Square"

client = L402Client()

# Step 1: Request → 402
print(f"Requesting video: '{prompt}'")
challenge = client.request("generate-video", {
    "prompt": prompt,
    "duration": 5,
})

if challenge.get("already_paid"):
    result = challenge["result"]
else:
    # Step 2: Pay the invoice
    print(f"\nInvoice to pay:\n{challenge['invoice']}\n")
    preimage = input("Preimage (hex): ").strip()

    # Step 3: Submit with proof
    print("\nSubmitting with L402 proof...")
    result = client.complete(challenge, preimage)

# Handle async response
body = result.get("body", result)

if isinstance(body, dict) and body.get("requestId"):
    request_id = body["requestId"]
    job_type = "video"
    print(f"\nJob submitted: {request_id}")
    print("Polling for completion...")

    while True:
        time.sleep(5)
        status = client.poll_job(request_id, job_type)
        state = status.get("status", "unknown")
        print(f"  Status: {state}")

        if state == "COMPLETED":
            video_url = status.get("result", {}).get("url", "")
            print(f"\nVideo ready: {video_url}")
            break
        elif state == "FAILED":
            print(f"\nJob failed: {status}")
            break
else:
    print(f"Response: {body}")

File: examples/node/generate-video.mjs
/**
 * Generate a video using Sats4AI L402 API (async service).
 *
 * Usage:
 *   node generate-video.mjs "A robot dancing in Times Square"
 *
 * Requires: Node.js 18+ (built-in fetch)
 */

import { requestL402, completeL402, pollJob } from "../../lib/l402-client.mjs";
import { createInterface } from "readline";

const prompt = process.argv[2] || "A robot dancing in Times Square";

console.log(`Requesting video: '${prompt}'`);

const body = { prompt, duration: 5 };

// Step 1: Request → 402
const challenge = await requestL402("generate-video", body);

let result;
if (challenge.alreadyPaid) {
  result = challenge.result;
} else {
  // Step 2: Pay the invoice
  console.log(`\nInvoice to pay:\n${challenge.invoice}\n`);

  const rl = createInterface({ input: process.stdin, output: process.stdout });
  const preimage = await new Promise((resolve) =>
    rl.question("Preimage (hex): ", (answer) => {
      rl.close();
      resolve(answer.trim());
    })
  );

  // Step 3: Submit with proof
  console.log("\nSubmitting with L402 proof...");
  result = await completeL402(challenge, body, preimage);
}

// Handle async response
const responseBody = result?.body || result;

if (responseBody?.requestId) {
  const { requestId } = responseBody;
  console.log(`\nJob submitted: ${requestId}`);
  console.log("Polling for completion...");

  while (true) {
    await new Promise((r) => setTimeout(r, 5000));
    const status = await pollJob(requestId, "video");
    console.log(`  Status: ${status.status}`);

    if (status.status === "COMPLETED") {
      console.log(`\nVideo ready: ${status.result?.url}`);
      break;
    } else if (status.status === "FAILED") {
      console.log(`\nJob failed:`, status);
      break;
    }
  }
} else {
  console.log("Response:", responseBody);
}
